### ISSUE TITLE
Weird recognition of type with generic mixin class

### ISSUE THREAD

**Issue by JensHeinrich** [microsoft/pyright#7400]
**Describe the bug**
When comparing the types of properties of a class to a protocol something weird happens, if that protocol is generic.

This might be similar to #6183

An additional test just showed me, that the error just might be in the matching of generic protocols.

**Code or Screenshots**

```python
from dataclasses import dataclass
from typing import TYPE_CHECKING
from typing import Generic
from typing import Protocol
from typing import TypeAlias
from typing import TypeVar

# Deliberately use multiple typevars to make if clear which is used where
A_co = TypeVar("A_co", covariant=True)
B_co = TypeVar("B_co", covariant=True)
C_co = TypeVar("C_co", covariant=True)
D_co = TypeVar("D_co", covariant=True)
E_co = TypeVar("E_co", covariant=True)
F_co = TypeVar("F_co", covariant=True)


class BProto(Protocol[A_co]):
    """Simple protocol."""

    @property
    def t(self) -> A_co: ...


@dataclass
class BModel(Generic[B_co]):
    """Simple dataclass fulfilling BProto."""

    t: B_co


if TYPE_CHECKING:
    b_model = BModel(t="t")
    b_model_type: BProto[str] = b_model
    specified_b_model = BModel[str](t="t")
    specified_b_model_type: BProto[str] = specified_b_model


class BRootProto(Protocol[D_co]):
    """Proto indicating `root` is BProto"""

    @property
    def root(self) -> BProto[D_co]: ...


class BRootMixin(Generic[E_co]):
    """Mixin to support nesting BProtos in `root`."""

    @property
    def t(self: BRootProto[E_co]) -> E_co:
        return self.root.t


@dataclass
class GenericManualModel(BRootMixin[F_co], Generic[F_co]):
    """Model using the mixin."""

    root: BModel[F_co]


if TYPE_CHECKING:
    generic_manual_model = GenericManualModel(root=b_model)
    # The error here is correct-ish, but could be more explicit
    generic_manual_model_type: BProto = generic_manual_model
    b_model_str = BModel[str](t="t")
    b_model_str_type: BProto[str] = b_model_str
    specified_generic_manual_model = GenericManualModel[str](root=b_model_str)
    attribute_type: str = specified_generic_manual_model.t
    # The error here shouldn't be one IMO
    specified_generic_manual_model_type: BProto[str] = specified_generic_manual_model


@dataclass
class ManualModel(BRootMixin[str]):
    """Simple model using the mixin."""

    root: BModel[str]


if TYPE_CHECKING:
    manual_model = ManualModel(root=b_model)
    # The error here is correct-ish, but could be more explicit
    manual_model_type: BProto = manual_model
    s_b_model_str = BModel[str](t="t")
    s_b_model_str_type: BProto[str] = s_b_model_str
    specified_manual_model = ManualModel(root=s_b_model_str)
    s_attribute_type: str = specified_manual_model.t
    # The error here shouldn't be one IMO
    specified_manual_model_type: BProto[str] = specified_manual_model

StrBRootMixin: TypeAlias = BRootMixin[str]


@dataclass
class AliasedManualModel(BRootMixin[str]):
    """Simple model using a TypeAlias of the mixin."""

    root: BModel[str]


if TYPE_CHECKING:
    aliased_manual_model = AliasedManualModel(root=b_model)
    # The error here is correct-ish, but could be more explicit
    manual_model_type: BProto = aliased_manual_model
    aliased_s_b_model_str = BModel[str](t="t")
    aliased_s_b_model_str_type: BProto[str] = aliased_s_b_model_str
    aliased_specified_manual_model = AliasedManualModel(root=s_b_model_str)
    aliased_s_attribute_type: str = aliased_specified_manual_model.t
    # The error here shouldn't be one IMO
    aliased_specified_manual_model_type: BProto[str] = aliased_specified_manual_model

# The following models work

class HardCodedBRootMixin:
    """Mixin to support nesting BProto[str] in `root`."""

    @property
    def t(self: BRootProto[str]) -> str:
        return self.root.t


@dataclass
class HardCodedManualModel(HardCodedBRootMixin):
    """Simple model using a hard coded version of the mixin."""

    root: BModel[str]


if TYPE_CHECKING:
    hard_coded_manual_model = HardCodedManualModel(root=b_model)
    # The error from above not being here is kinda weird as we are comparing a property of type `str` to a property of type `Unknown`
    manual_model_type: BProto = hard_coded_manual_model
    hard_coded_s_b_model_str = BModel[str](t="t")
    hard_coded_s_b_model_str_type: BProto[str] = hard_coded_s_b_model_str
    hard_coded_specified_manual_model = HardCodedManualModel(root=s_b_model_str)
    hard_coded_s_attribute_type: str = hard_coded_specified_manual_model.t
    # This is correct now
    hard_coded_specified_manual_model_type: BProto[str] = (
        hard_coded_specified_manual_model
    )


```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
```sh
> poetry run pyright --version
pyright 1.1.352
```

**Comment by JensHeinrich** [comment]
[pyright-playground](https://pyright-play.net/?strict=true&locale=en&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sBYAKFEigQE8j8A7AcwONPIAqATQAKAUQD6AYQASYqQGkAkgDkA4vUbRW7brxJkoa7B2wh8yTeG1tOPQgfIjwCMMjCorTHXf38oAmzYAIKo%2BJheNrr2fIaBRNgAanAg9PQAxFAAIthhAEZmiLksUACuOFAQpagI%2BESo2MxBAG4peK6VcADWjfjAUGjYKVAA7gAWFmMEeOXYMKNjZtj0wRLuUAC8AUHJIAAUAESr7gcANANgreZwHAgbAiCl2ACU9ABCa2Cb2wm7hx8nc7uK7hW73R4vehST7feJJFKHaGAi4gm53B5PV50LIwrZwv4HHHI4EpUHoiFYyTrPE7BEHKlgM4o0lo8GY%2BgAMVxP3h%2BwOXOJlxZYIxkLoaToaEweDezjArj2ctc7lQAG1jmAALrPABc9CgBqgB2NAGUHA0oEQXG4PAA6Y0HCWGqAAAStYASZBY%2BsNMGw-QQexwqGAzygAFoAHxQDU6qC2hMS%2Bgu%2BBIKW0SXUGUAWTAftQe2MpnMyFVAK1up9BodZuIFtTVGlUGA1WA%2BFQYT0sut9uNTsNCDj5aTdD6AVEklk8mU6j1dGdeQkEDzuW%2Bb1z%2Bb2dwOCAOWIXS5XqAkOmwQ6VYFVGAQIE130Xy-zVagGASyD6%2BDmEgfR7XG9yV43pqW4bDue7Pq%2B2Dvm2X4-vmJ5BOe1qAbe3yQdBn4wN%2Bh5PuKeHpjKABKYDyheirWiqqpEhWc7Og6F4EBwMAWIgvgAAbgPKbHTFA3byoyfZ4c6brgJ6rDPn6-ScYGwahhG0Z8a4VGfJqcYJraI4EbxxHytm%2BAAB6cIWJhmBYqoMtqtGGg6emGRwzBfBgpREI4UCmNeviKWAeCcFAHEkQgbG9o6QmGiJHpmOJ86%2Bv6zBBrkwBDjpCAXuZKlhlGUAMlZzoGiA2AIKUID2bJtrSbaCAjimlAEfQWlFqZyDZjcpRwKg-4Fm8yW2ZwqoClq5wNSWfXpTlRrGh1ZQYL4CCLJUBmcMF-Z5QFQ4dSN7iaiOY7COI0hyIoqhqGNXAmSWS4tW1OGrlsQ0WM1HCte1R57NJGxwbk%2B6GpkAhzWY4AgFAiz5Tx7ggPlyAIOG%2BAYGM5x5KU5DuNU8wFJUpCNNg%2Bn1BY%2BCVdFBqncWFgXY9V2PrkCEJEh-HfETjWk0912eATUAfce16A1s65HihwHbruX0GuzEic1TZ68alnN3lsIucxBb4fl%2B9PnRAl3HhTqDfHdTXq%2Bt0uvQF73M6LN5C1AiA3vgCMINg4txpzaGKzBWEqyTatkxrR4Vc%2BP1-eDpBA0sL5jGAKMcAA5OQaNgKYUBKNmADyCtQUrrtne76sm6eNNKdLTupy7Ehu8gjPk0eVUNrVmZNg9T0dXsXUBT1HB85WrM1uajSa1NM1zRAC0cEtoUrfKa289L239Ltk4HTOx3Ph7TM91sddtQ3b3s%2BbfuYwHgPA70eBgxDUMw3DbOIxcKNs93GNQFjOPvvjzpL%2BX8E55L1rfK-Xu4c6GDYU1qbLmvF9ZARAmBc2AC5Y3ntp-fifM0KAKPMAlOGEvw-2Zt8Nez1NxvWgSbTmUCJCW3MDbO2H9HZbHQmnMuv9cg%2B1Zjve%2Be8g4g1hmHVAMBI7R0aLHRoCdk6sxoUXTBQCP5eUQdQ52mE6HMwlCaG8TddKDzjHCUI4Q8Dc26oPPmlcapZjqlmGMYRMBzBwQ3ZRCAW5tzGp3Ost98y9z0HAHkGjMBQDAAGfug9h7PmkuPfMei8L0B2hOfa04jpjTapojBWcV4mNiTACxL1N7M23gEf2AM2GHwuODKCp9Ybw0vsjLhN90YgwfmEJ%2Bi94koIkQxLYMSzFYTERXVmzScBYQIUAqhoCJ7gIFuBDppiumi2QfBMWDTkL5yaaMr8PSUHyxGUk0WMi4meywVsdxXSUl4KNosyZZtnydIWSQhAVtyFwL6ac7p6zWl1PzIw50zD-qBwPiHTh3Co7lP4fHJOJz5l3MLrItp79ELwLzkBb4ty1kgo2cvdpEotIyBSDAKQK4YBWJbnYiag8HIvmcq5dytQuxS2hb5fyXE-Gs3CmJb0rNJJxVkklAK5LbwZWjJzMazp8qFWKi%2BBKZUAqMIlNVNMhia5YCgKikA6LMV7NyHsWV8q-RYp0XZdudFTRd3Rk48ovhXFjDRRcNVUBmhmGmrHTx3ju6%2BIdMtKAAT%2BlBMniE0c09wlTkOrOZ8xq5WfDVXIhJKqMVqsVQWNJmsMm-V3tkrQFs8iXEaBweUN9fAfJhlALonB4CjGwPgOVtTNniIhV5b4-qsLuCDWC3Ifq0WBrOTAkBPNXWDNAoLetAbq1NsIbA6ZCDZlAwbT27pEzKbLOdJWxtwL0EPJLb%2BLYoaFV61SQc8dHNjms2naO8ZpDraIwoRCvpO7MVwrncG72z9vqZKzVm4%2BBS3JgBGF2qtZ6RGgseZTAdULUJbD2M%2BKdI7333MvX-Q0WIgA)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.355, which I just published.
